### ISSUE TITLE
Ability to add extra parameters on creation of NAV Container

### ISSUE THREAD

**Issue by nickrobbo** [NAVBaaS/GitIntegration#3]
Would it be possible to add some functionality to copy Add-Ins to the container? Or specify any extra parameters, like the examples in this [link ](https://blogs.msdn.microsoft.com/freddyk/2018/03/22/navcontainerhelper-setup-cside-development-environment-with-source-code-management/)

```
New-NavContainer -accept_eula `
                 -containerName $containerName `
                 -imageName $imageName `
                 -auth Windows `
                 -licensefile $mylicense `
                 -updateHosts `
                 -includeCSide `
                 -additionalParameters @("--volume ${sourceFolder}:c:\source",
                                         "--volume ${addInsFolder}:c:\run\Add-Ins",
                                         "--env CustomNavSettings=EnableTaskScheduler=true")
```

**Comment by richardrobberse** [comment]
Addins:
I was waiting for this one, but I'm still thinking of a way to handle this in a nice way because I don't want to bother people with all kind of complex scenario's while they just want a basic container.
You can workaround this for know by copying the dlls to the container's add-ins folder.

Volumes:
When creating the container there's already a folder mapped to your container, it's somewhere in your temp directory.
You can find it by using the Get-NavContainerSharedFolders cmdlet.

CustomNavSettings:
I'm not sure (yet) if I want to support this one, I know that for example the Task Scheduler is enabled by default.

**Comment by nickrobbo** [comment]
right OK, thanks for the swift reply.  I have already created some scripts to import our base objects as a fob and compile them, so I will just create another function to copy our DLL's to the correct folder.

**Comment by richardrobberse** [comment]
Would it help if would add a setting where you can store the path to a certain PowerShell script, and in that script you can then populate the additional parameters which I will take into account when creating the container?

**Comment by nickrobbo** [comment]
Yes I guess that would help, an example might help me understand how that could be put into practice? As in how we would specify these parameters in the PowersShell Script?

**Comment by richardrobberse** [comment]
You could then create a script like this:
$SourceFolder = "c:\temp"
$AddInsFolder = "c:\addins"

$AdditionalParams = @()
$AdditionalParams += "--volume ${sourceFolder}:c:\source"
$AdditionalParams += "--volume ${AddInsFolder}:c:\run\Add-Ins"
$AdditionalParams += "--env CustomNavSettings=EnableTaskScheduler=true"

Which I will call before creating the container.
The only downside is that you're not flexible and the settings are hard-coded.

Another option is that I call a script that you specify through a path after creating the container where you can basically do anything you want except mapping folders because that's not possible afterwards.

**Comment by nickrobbo** [comment]
OK, in my scripts I reference a setup.json file to put the names of the db, server and file locations, I could just add a property to specify the directory where my add-ins are located.  Then get those values and pass them as additional parameters.

**Comment by richardrobberse** [comment]
My current idea is to natively support add-ins - by for example specifying the folder during the Create Container command.

Next to that, I want to add the ability to hook-in before creating the container and after creating the container.
I'll add two settings where you can specify the path to the PowerShell script that should be executed:
OnBeforeCreateContainer
OnAfterCreateContainer

I will also update the documentation with the global variables that are available in those scripts.

Those scripts will run on the host - so not inside the container.
You should then add those settings yourself in your user settings.

Agree?

**Comment by nickrobbo** [comment]
This sounds great, will out of the box give us support for Add-Ins. Plus some extra flexibility via those hooks.

**Comment by richardrobberse** [comment]
Nice, I've just implemented the two hooks.
I might release that before actually starting with the native add-ins support so that you can test with it for a bit.

I'm also thinking to reduce the number of information messages popping up in VS Code and only show one information message whenever the called command (for example Sync) is finished.
What's your opinion about this?

**Comment by richardrobberse** [comment]
Hooks are implemented in version 1.2.0 which released last night.
I will create a new issue for the out of the box custom add-in support.

**Comment by nickrobbo** [comment]
Brilliant, I will download the update and give it a go when I have some spare time later this week.  Thanks for the update!
